### PR TITLE
Feature tmc#189 error success ignore states

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: sdl2promela/stop_condition/translator.py [hunk 1]
```diff
@@ -1099,6 +1131,45 @@ def _add_precheck_expressions(
...
+) -> promela.Expression:
+    def joiner(lhs: promela.Expression, rhs: promela.Expression) -> promela.Expression:
+        return (
+            promelaBuilder.BinaryExpressionBuilder(op)
+            .withLeft(lhs)
+            .withRight(rhs)
+            .build()
+        )
+
+    print("len ", len(expressions))
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/translator.py:1145]: ?

File: sdl2promela/stop_condition/translator.py [hunk 2]
```diff
@@ -1112,6 +1183,29 @@ def _generate_filter_out_alternative(
         expression = _generate(context, statements[index].expression)
         expression = _add_precheck_expressions(context, expression)
         expressions.append(expression)
+    # For all observers with ignore state generate list of expressions
+    for observer in context.observers_with_ignore:
+        ignorestates = context.processes[observer].ignorestates
+        expression = _create_observer_state_check(ignorestates, observer)
+        expressions.append(expression)
+
+    if context.observers_with_success:
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/translator.py:1192]: I don't understand why it is here. Maybe the host function should receive a comment?

File: sdl2promela/stop_condition/translator.py [hunk 3]
```diff
@@ -1112,6 +1183,29 @@ def _generate_filter_out_alternative(
...
+        # Generate one expression AND for all observers with success state
+        # Join generated expression using AND
+        # Append result to list of expression
+        all_observer_expressions: List[promela.Expression] = []
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/translator.py:1196]: This code is repeated, right? Can it be extracted?

File: sdl2promela/stop_condition/translator.py [hunk 4]
```diff
@@ -1234,6 +1364,20 @@ def _translate_basic_statements(
...
         builder.withAlternative(_generate_never_alternative(never, context))
 
+    if context.observers_with_error:
+        for observer in context.observers_with_error:
+            states = context.processes[observer].errorstates
+            builder.withAlternative(
+                _create_never_alternative_for_observer(states, observer)
+            )
+
+    if context.search_for_observer_success and context.observers_with_success:
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/translator.py:1374]: This requires a comment for future generations

File: sdl2promela/translator.py [hunk 5]
```diff
@@ -2087,6 +2088,24 @@ def __generate_aggregate_inline(
     return builder.build()
 
 
+def __generate_observer_variables(context: Context, builder: ModelBuilder):
+    if context.sdl_model.errorstates:
```
Comment (Lurkerpas) [inline sdl2promela/translator.py:2092]: Why is this needed?

File: sdl2promela/sdl2promela.py [hunk 6]
```diff
@@ -333,9 +364,12 @@ def main():
     if arguments.verbose:
         __log.setLevel(level=logging.INFO)
 
-    if arguments.scl_files:
+    if arguments.scl_files or arguments.include_observers:
```
[the remark this answers is not in the document]
Comment (rphh) [reply sdl2promela/sdl2promela.py:367]: then no checks will be generated. This argument was generated for the scenario without stop conditions and with observers. In such case, the never claim shall be generated.

File: sdl2promela/sdl2promela.py [hunk 7]
```diff
@@ -27,15 +27,21 @@ class ProgramOptions:
     """Show verbose output."""
     sdl_files: List[List[str]]
     """Groups of sdl files, every group shall contains one process."""
```
[the remark this answers is not in the document]
Comment (rphh) [reply sdl2promela/sdl2promela.py]: changed

File: sdl2promela/stop_condition/translator.py (hunk 1 above)
Comment (rphh) [reply sdl2promela/stop_condition/translator.py]: removed (It was used for development)

File: sdl2promela/stop_condition/translator.py (hunk 2 above)
Comment (rphh) [reply sdl2promela/stop_condition/translator.py:1190]: the comment below was extended

File: sdl2promela/stop_condition/translator.py (hunk 4 above)
Comment (rphh) [reply sdl2promela/stop_condition/translator.py:1375]: added a comment block

File: sdl2promela/translator.py (hunk 5 above)
Comment (rphh) [reply sdl2promela/translator.py]: early idea, removed

File: sdl2promela/stop_condition/translator.py (hunk 3 above)
Comment (rphh) [reply sdl2promela/stop_condition/translator.py:1197]: Hmm, there's similar code, which wraps the expression using unary NOT, so there are two versions used in two places. I do not see a need to extract and parametrize it now, maybe in future

**Review by Lurkerpas — APPROVED** [review]
